File: integration/BattleCityDemo/Assets/Scripts/behaviac/BehaviorTree/Nodes/Actions/Assignment.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections;
using System.Collections.Generic;

namespace behaviac
{
    public class Assignment : BehaviorNode
    {
        public Assignment()
        {
		}
        ~Assignment()
        {
            m_opl = null;
            m_opr = null;
            m_opr_m = null;
        }

        protected override void load(int version, string agentType, List<property_t> properties)
        {
            base.load(version, agentType, properties);

            string propertyName = null;

            foreach (property_t p in properties)
            {
                if (p.name == "Opl")
                {
                    this.m_opl = Condition.LoadLeft(p.value, ref propertyName, null);
                }
                else if (p.name == "Opr")
                {
                    int pParenthesis = p.value.IndexOf('(');
                    if (pParenthesis == -1)
                    {
                        string typeName = null;
                        this.m_opr = Condition.LoadRight(p.value, propertyName, ref typeName);
                    }
                    else
                    {
                        //method
                        this.m_opr_m = Action.LoadMethod(p.value);
                    }
                }
                else
                {
                    //Debug.Check(0, "unrecognised property %s", p.name);
                }
            }
        }

        public override bool IsValid(Agent pAgent, BehaviorTask pTask)
        {
            if (!(pTask.GetNode() is Assignment))
            {
                return false;
            }

            return base.IsValid(pAgent, pTask);
        }

        protected override BehaviorTask createTask()
        {
            return new AssignmentTask();
        }

        protected Property m_opl;
        protected Property m_opr;
        protected CMethodBase m_opr_m;

        class AssignmentTask : LeafTask
        {
            public AssignmentTask()
            { }

            ~AssignmentTask()
            {

            }

            public override void copyto(BehaviorTask target)
            {
                base.copyto(target);
            }

            public override void save(ISerializableNode node)
            {
                base.save(node);
            }
            public override void load(ISerializableNode node)
            {
                base.load(node);
            }

            protected override bool isContinueTicking()
            {
                return false;
            }

            protected override bool onenter(Agent pAgent)
            {
                return true;
            }
            protected override void onexit(Agent pAgent, EBTStatus s)
            {
            }

            protected override EBTStatus update(Agent pAgent, EBTStatus childStatus)
            {
                EBTStatus result = EBTStatus.BT_SUCCESS;

                Debug.Check(this.GetNode() is Assignment);
                Assignment pAssignmentNode = (Assignment)(this.GetNode());

                if (pAssignmentNode.m_opr_m != null)
                {
                    ParentType pt = pAssignmentNode.m_opr_m.GetParentType();
                    Agent pParent = pAgent;
                    if (pt == ParentType.PT_INSTANCE)
                    {
                        pParent = Agent.GetInstance(pAssignmentNode.m_opr_m.GetInstanceNameString(), pParent.GetContextId());
						Debug.Check(pParent != null || Utils.IsStaticClass(pAssignmentNode.m_opr_m.GetInstanceNameString()));
                    }

                    object returnValue = pAssignmentNode.m_opr_m.run(pParent, pAgent);

                    ParentType pt_opl = pAssignmentNode.m_opl.GetParentType();
                    Agent pParentOpl = pAgent;
                    if (pt_opl == ParentType.PT_INSTANCE)
                    {
                        pParentOpl = Agent.GetInstance(pAssignmentNode.m_opl.GetInstanceNameString(), pParentOpl.GetContextId());
						Debug.Check(pParentOpl != null || Utils.IsStaticClass(pAssignmentNode.m_opl.GetInstanceNameString()));
                    }

                    pAssignmentNode.m_opl.SetValue(pParentOpl, returnValue);
                }
                else if (pAssignmentNode.m_opr != null && pAssignmentNode.m_opl != null)
                {
                    Agent pParentL = pAgent;
                    Agent pParentR = pAgent;

                    {
                        ParentType pt = pAssignmentNode.m_opl.GetParentType();
                        if (pt == ParentType.PT_INSTANCE)
                        {
                            pParentL = Agent.GetInstance(pAssignmentNode.m_opl.GetInstanceNameString(), pParentL.GetContextId());
							Debug.Check(pParentL != null || Utils.IsStaticClass(pAssignmentNode.m_opl.GetInstanceNameString()));
                        }
                    }
                    {
                        ParentType pt = pAssignmentNode.m_opr.GetParentType();
                        if (pt == ParentType.PT_INSTANCE)
                        {
                            pParentR = Agent.GetInstance(pAssignmentNode.m_opr.GetInstanceNameString(), pParentR.GetContextId());

                            //it is a const
                            if (pParentR == null)
                            {
                                pParentR = pParentL;
                            }
                        }
                    }

                    pAssignmentNode.m_opl.SetFrom(pParentR, pAssignmentNode.m_opr, pParentL);
                }
                else
                {
                    result = pAssignmentNode.update_impl(pAgent, childStatus);
                }

                return result;
            }
        }
    }
}

File: src/behaviortree/nodes/actions/assignment.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/base/base.h"
#include "behaviac/property/property_t.h"
#include "behaviac/behaviortree/nodes/actions/assignment.h"
#include "behaviac/base/core/profiler/profiler.h"
#include "behaviac/base/object/method.h"

#include "behaviac/behaviortree/nodes/actions/action.h"
#include "behaviac/behaviortree/nodes/conditions/condition.h"

namespace behaviac
{
	Assignment::Assignment() : m_opl(0), m_opr(0), m_opr_m(0), m_bCast(false)
    {
    }

    Assignment::~Assignment()
    {
        BEHAVIAC_DELETE(m_opr_m);
    }

    //behaviac::CMethodBase* LoadMethod(const char* value);
    //Property* LoadLeft(const char* value, behaviac::string& propertyName, const char* constValue);
    //Property* LoadRight(const char* value, const behaviac::string& propertyName, behaviac::string& typeName);
    /**
    handle the Assignment property

    */
    void Assignment::load(int version, const char* agentType, const properties_t& properties)
    {
        super::load(version, agentType, properties);

        behaviac::string propertyName;

        for (propertie_const_iterator_t it = properties.begin(); it != properties.end(); ++it)
        {
            const property_t& p = (*it);

			if (StringUtils::StrEqual(p.name, "CastRight"))
			{
				this->m_bCast = StringUtils::StrEqualNoCase(p.value, "true");
			}
			else if (StringUtils::StrEqual(p.name, "Opl"))
            {
				this->m_opl = Condition::LoadLeft(p.value, propertyName);
            }
			else if (StringUtils::StrEqual(p.name, "Opr"))
            {
                const char* pParenthesis = strchr(p.value, '(');

                if (pParenthesis == 0)
                {
                    behaviac::string typeName;
                    this->m_opr = Condition::LoadRight(p.value, typeName);
                }
                else
                {
                    //method
                    this->m_opr_m = Action::LoadMethod(p.value);
                }
            }
            else
            {
                //BEHAVIAC_ASSERT(0, "unrecognised property %s", p.name);
            }
        }
    }
    bool Assignment::EvaluteAssignment(bool bCast, const Agent* pAgent, Property* opl, Property* opr, behaviac::CMethodBase* opr_m)
    {
        bool bValid = false;

		if (opl != NULL)
		{
			if (opr_m != NULL)
			{
				Agent* pParentL = (Agent*)opl->GetParentAgent(pAgent);

				opl->SetFrom((Agent*)pAgent, opr_m, pParentL, bCast);

				bValid = true;
			}
			else if (opr != NULL)
			{
				Agent* pParentL = opl->GetParentAgent(pAgent);
				Agent* pParentR = opr->GetParentAgent(pAgent);

				opl->SetFrom(pParentR, opr, pParentL, bCast);

				bValid = true;
			}
		}

        return bValid;
    }
    bool Assignment::IsValid(Agent* pAgent, BehaviorTask* pTask) const
    {
        if (!Assignment::DynamicCast(pTask->GetNode()))
        {
            return false;
        }

        return super::IsValid(pAgent, pTask);
    }

    BehaviorTask* Assignment::createTask() const
    {
        AssignmentTask* pTask = BEHAVIAC_NEW AssignmentTask();

        return pTask;
    }

    AssignmentTask::AssignmentTask() : LeafTask()
    {
    }

    AssignmentTask::~AssignmentTask()
    {
    }

    void AssignmentTask::copyto(BehaviorTask* target) const
    {
        super::copyto(target);
    }

    void AssignmentTask::save(ISerializableNode* node) const
    {
        super::save(node);
    }

    void AssignmentTask::load(ISerializableNode* node)
    {
        super::load(node);
    }


    bool AssignmentTask::onenter(Agent* pAgent)
    {
        BEHAVIAC_UNUSED_VAR(pAgent);
        return true;
    }

    void AssignmentTask::onexit(Agent* pAgent, EBTStatus s)
    {
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_UNUSED_VAR(s);
    }

    EBTStatus AssignmentTask::update(Agent* pAgent, EBTStatus childStatus)
    {
        BEHAVIAC_ASSERT(childStatus == BT_RUNNING);

        BEHAVIAC_ASSERT(Assignment::DynamicCast(this->GetNode()));
        Assignment* pAssignmentNode = (Assignment*)(this->GetNode());

        EBTStatus result = BT_SUCCESS;
		bool bValid = Assignment::EvaluteAssignment(pAssignmentNode->m_bCast, pAgent, pAssignmentNode->m_opl, pAssignmentNode->m_opr, pAssignmentNode->m_opr_m);

        if (!bValid)
        {
            result = pAssignmentNode->update_impl(pAgent, childStatus);
        }

        return result;
    }
}
